Source/WebCore/style/ElementRuleCollector.cpp

```diff
@@ -602,6 +619,8 @@ void ElementRuleCollector::collectMatchingRulesForList(const RuleSet::RuleDataVe
         }
 
         addRuleIfMatches();
+        if (isEarlyReturnModeAndHasMatchedAnyRules())
+            return;
```

Comment:
We still do some unnecessary work like hash lookups in `ElementRuleCollector::collectMatchingRules` after the first match. I think this ok for now though, it can be improved later.